File: Week Four/Find Bottom Left Tree Value.js
var findBottomLeftValue = function (root) {
  if (!root) {
    return null;
  }
  let dist = 0;
  return helper(root, dist).node.val;
};

var helper = (root, dist) => {
  if (!root) {
    return null;
  }
  if (!root.left && !root.right) {
    return { node: root, length: dist };
  }
  if (root.left && !root.right) {
    let left = helper(root.left, dist + 1);
    return left;
  }
  if (!root.left && root.right) {
    let right = helper(root.right, dist + 1);
    return right;
  }
  let left = helper(root.left, dist + 1);
  let right = helper(root.right, dist + 1);

  console.log(left, right);
  return left.length >= right.length ? left : right;
};

File: Week Five/Find Bottom Left Tree Value.py
class Solution:
    def findBottomLeftValue(self, root) -> int:
        return self.helper(root)[0]

    def helper(self, root):
        if not root:
            return (0, 0)

        if not root.left and not root.right:
            return (root.val, 1)

        left = self.helper(root.left)
        right = self.helper(root.right)

        if left[1] < right[1]:
            return (right[0], right[1]+1)
        else:
            return (left[0], left[1]+1)
